File: src/main/java/me/coderfrish/toml/reader/TomlLexer.java
package me.coderfrish.toml.reader;

import me.coderfrish.toml.reader.token.Token;
import me.coderfrish.toml.reader.token.TokenType;

import java.util.ArrayList;
import java.util.List;

public class TomlLexer {
    private static final String WORD_PATTERN = "[A-Za-z]";
    private static final String NUMBER_PATTERN = "[0-9]";
    private final char[] chars;
    private int offset = 0;

    public TomlLexer(String toml) {
        this.chars = toml.toCharArray();
    }

    public List<Token> tokenize() {
        final List<Token>  tokens = new ArrayList<>();

        while (offset < chars.length) {
            char current = chars[offset];

            if (current == '\n') {
                offset++;
                continue;
            }

            if (matches(current, WORD_PATTERN)) {
                StringBuilder builder = new StringBuilder();
                while (offset < chars.length && matches(chars[offset], WORD_PATTERN)) {
                    builder.append(chars[offset]);
                    offset++;
                }
                tokens.add(new Token(TokenType.WORD, builder));
                continue;
            } else if (matches(current, NUMBER_PATTERN)) {
                StringBuilder builder = new StringBuilder();
                while (offset < chars.length && matches(chars[offset], NUMBER_PATTERN)) {
                    builder.append(chars[offset]);
                    offset++;
                }
                tokens.add(new Token(TokenType.NUMBER, builder));
                continue;
            } else {
                tokens.add(new Token(TokenType.SYMBOL, String.valueOf(current)));
            }

            offset++;
        }

        return tokens;
    }

    private static boolean matches(char c, String pattern) {
        return String.valueOf(c).matches(pattern);
    }
}

File: src/main/kotlin/me/coderfrish/toml/reader/TomlLexer.kt
package me.coderfrish.toml.reader

import me.coderfrish.toml.base.BaseLexer
import me.coderfrish.toml.tokens.TokenType
import me.coderfrish.toml.tokens.TomlToken
import me.coderfrish.toml.shared.BOOLEAN_PATTERN
import me.coderfrish.toml.shared.IDENTIFIER_PATTERN
import me.coderfrish.toml.shared.MatchUtils.matches
import me.coderfrish.toml.tokens.TokenType.STRING
import me.coderfrish.toml.tokens.TokenType.NUMBER
import me.coderfrish.toml.tokens.TokenType.EQUALS
import me.coderfrish.toml.tokens.TokenType.BOOLEAN
import me.coderfrish.toml.tokens.TokenType.IDENTIFIER
import me.coderfrish.toml.shared.MatchUtils.isRawNumber
import me.coderfrish.toml.shared.QUOTA_NOT_CONSISTENT_EXCEPTION


/**
 * @author Frish2021
 */
class TomlLexer(str: String) : BaseLexer(str) {
    override fun parseIdentifier(): TomlToken {
        val builder = StringBuilder()
        while (offset < str.length && matches(peek(), IDENTIFIER_PATTERN)) {
            builder.append(peekWithAdvance())
        }

        return parseIdentifierToken(builder.toString())
    }

    private fun parseIdentifierToken(string: String): TomlToken =
        TomlToken(parseIdentifierType(string), string)

    private fun parseIdentifierType(string: String): TokenType =
        if (isNotTokenEmpty() && lastType().equals(EQUALS)) {
            if (matches(string, BOOLEAN_PATTERN))
                BOOLEAN
            else if (isRawNumber(string))
                NUMBER
            else
                IDENTIFIER
        } else {
            IDENTIFIER
        }

    /* 除了解析字符串那边是一大坨其他都还行把... */
    override fun parseString(char: Char): TomlToken {
        this.advance()
        val builder = StringBuilder()
        while (offset < str.length && (peek() != '"' && peek() != '\'')) {
            builder.append(peekWithAdvance())
        }
        if (peek() != char) error(QUOTA_NOT_CONSISTENT_EXCEPTION)
        this.advance()

        return parseStringToken(builder.toString())
    }

    private fun parseStringToken(string: String): TomlToken =
        TomlToken(parseStringType(), string)

    private fun parseStringType(): TokenType =
        if (isNotTokenEmpty() && lastType().equals(EQUALS)) {
            STRING
        } else {
            IDENTIFIER
        }
}
